src/ripple/app/tx/impl/SusPay.cpp

```diff
@@ -264,45 +291,61 @@ SusPayFinish::preflight (PreflightContext const& ctx)
     if (!isTesSuccess (ret))
         return ret;
 
-    if (ctx.tx[~sfMethod])
+    auto const cb = ctx.tx[~sfCondition];
+    auto const fb = ctx.tx[~sfFulfillment];
+
+    // If you specify a condition, then you must specify
+    // a fulfillment.
+    if (static_cast<bool>(cb) != static_cast<bool>(fb))
+        return temMALFORMED;
+
+    if (cb && fb)
     {
-        // Condition
-        switch(ctx.tx[sfMethod])
-        {
-        case 1:
+        using namespace ripple::cryptoconditions;
+
+        // TODO: Remove this try/catch once cryptoconditions
+        //       no longer use exceptions.
+
+        // CHECKME: Should we consider doing this in doApply?
+        //          Yes, cryptographic operations are involved
+        //          but we could charge a fee when the specified
+        //          condition or fulfillment are wrong.
+        try
         {
-            if (! ctx.tx[~sfDigest])
-                return temMALFORMED;
-            if (! ctx.tx[~sfProof])
+            auto condition = loadCondition(*cb);
+
+            if (!condition)
                 return temMALFORMED;
-            if (ctx.tx[~sfProof]->size() != 32)
+
+            auto fulfillment = loadFulfillment(*fb);
+
+            if (!fulfillment)
                 return temMALFORMED;
-            sha256_hasher h;
-            using beast::hash_append;
-            hash_append(h, ctx.tx[sfProof]);
-            uint256 digest;
+
             {
-                auto const result = static_cast<
-                    sha256_hasher::result_type>(h);
-                std::memcpy(digest.data(),
-                    result.data(), result.size());
+                // TODO: This is here temporarily to ensure
+                //       that the condition & fulfillment
+                //       given don't contain unnecessary
+                //       trailing junk. The new parsing API
+                //       will simplify the checking here.
+
+                auto b1 = to_blob(*condition);
+                if (*cb != makeSlice(b1))
+                    return temMALFORMED;
+
+                auto b2 = to_blob(*fulfillment);
+                if (*fb != makeSlice(b2))
+                    return temMALFORMED;
```

Comment:
You can't easily put it into the `loadFulfillment`: consider the case where you're loading a threshold which contains multiple subfulfillments. It's OK to have stuff "left over" if you're parsing a subfulfillment (since you'll be getting to it shortly), but not if you're parsing a top-level fulfillment.

You'd either need two functions or a function that takes an optional `bool` or `int` argument indicating whether this is the outer fulfillment.